tx/textual/valuerenderer/coins.go

```diff
@@ -76,7 +83,120 @@ func (vr coinsValueRenderer) Format(ctx context.Context, v protoreflect.Value) (
 	}
 }
 
-func (vr coinsValueRenderer) Parse(_ context.Context, screens []Screen) (protoreflect.Value, error) {
-	// ref: https://github.com/cosmos/cosmos-sdk/issues/13153
-	panic("implement me, see #13153")
+func (vr coinsValueRenderer) Parse(ctx context.Context, screens []Screen) (protoreflect.Value, error) {
+	if len(screens) != 1 {
+		return nilValue, fmt.Errorf("expected single screen: %v", screens)
+	}
+
+	if screens[0].Text == emptyCoins {
+		if vr.fd != nil && vr.fd.IsList() {
+			return protoreflect.ValueOfList(listpb.NewGenericList([]*basev1beta1.Coin{})), nil
+		}
+
+		return protoreflect.ValueOfMessage((&basev1beta1.Coin{}).ProtoReflect()), nil
+	}
+
+	coins := strings.Split(screens[0].Text, ", ")
+	metadatas := make([]*bankv1beta1.Metadata, len(coins))
+
+	var err error
+	for i, coin := range coins {
+		coinArr := strings.Split(coin, " ")
+		if len(coinArr) != 2 {
+			return nilValue, fmt.Errorf("invalid coin %s", coins)
+		}
+		metadatas[i], err = vr.coinMetadataQuerier(ctx, coinArr[1])
+		if err != nil {
+			return nilValue, err
+		}
+	}
+
+	parsed, err := parseCoins(coins, metadatas)
+	if err != nil {
+		return nilValue, err
+	}
+
+	if vr.fd != nil && vr.fd.IsList() {
+		return protoreflect.ValueOf(listpb.NewGenericList(parsed)), err
+	} else {
+		return protoreflect.ValueOfMessage(parsed[0].ProtoReflect()), err
+	}
+}
+
+func parseCoins(coins []string, metadata []*bankv1beta1.Metadata) ([]*basev1beta1.Coin, error) {
+	if len(coins) != len(metadata) {
+		return []*basev1beta1.Coin{}, fmt.Errorf("formatCoins expect one metadata for each coin; expected %d, got %d", len(coins), len(metadata))
+	}
+
+	parsedCoins := make([]*basev1beta1.Coin, len(coins))
+	for i, coinStr := range coins {
+		coin, err := parseCoin(coinStr, metadata[i])
+		if err != nil {
+			return []*basev1beta1.Coin{}, err
+		}
+		parsedCoins[i] = coin
+	}
+
+	return parsedCoins, nil
+}
+
+// parseCoin parses a single value-rendered coin into the Coin struct.
+// It shares a lot of code with `cosmos-sdk.io/core/coins.Format`,
+// so this code might be refactored once we have
+// a core Parse function for coins.
+func parseCoin(coinStr string, metadata *bankv1beta1.Metadata) (*basev1beta1.Coin, error) {
+	coinArr := strings.Split(coinStr, " ")
+	amt1 := coinArr[0]
+	coinDenom := coinArr[1]
+
+	if metadata == nil || metadata.Base == "" || coinArr[1] == metadata.Base {
+		dec, err := parseDec(amt1)
+		return &basev1beta1.Coin{
+			Amount: dec,
+			Denom:  coinDenom,
+		}, err
```

Comment:
Why are we returning a non-nil coin and potentially a non-nil error? Let's be explicit here and do
```go
dec, err := parseDec(amt1)
if err != nil {
      return nil, err
}
return &basev1beta1.Coin{Amount: dec, Denom: coinDenom}, nil
```